File: src/engine/assembly.py
"""
拼装引擎：将用户已选条目列表 + 全局配置 → PromptIR + 最终字符串。

处理流程：
1. 规则校验与消解（调用 rules.apply_rules）
2. 排序（按维度 sort_order 或用户自定义顺序）
3. 生成 PromptIR
4. 适配为最终字符串（调用 adapters.adapt_to_model）
"""
from __future__ import annotations
from engine.models import (
    SelectedItem, AssemblyConfig, PromptIR, IRSegment, Module
)
from engine.rules import apply_rules
from engine.adapters import adapt_to_model


def assemble(
    selected: list[SelectedItem],
    config: AssemblyConfig,
) -> tuple[PromptIR, str]:
    """拼装主函数。

    Args:
        selected: 用户已选条目列表
        config: 拼装全局配置

    Returns:
        (ir, final_prompt) — 中间表示与最终字符串
    """
    # 1. 规则校验与消解
    filtered, warnings = apply_rules(selected)

    # 2. 排序
    ordered = sort_by_order(filtered, config.sort_by)

    # 3. 生成 IR
    ir = PromptIR(
        segments=[
            IRSegment(
                dimension_key=it.module.dimension_key or "",
                text=it.module.content_en,
                weight=it.weight_override if it.weight_override is not None else it.module.weight,
                source_module_id=it.module.id,
            )
            for it in ordered
        ],
        warnings=warnings,
    )

    # 4. 适配为最终字符串
    final_prompt = adapt_to_model(ir, config.model_profile, config)
    return ir, final_prompt


def sort_by_order(
    items: list[SelectedItem],
    mode: str,
) -> list[SelectedItem]:
    """排序已选条目。

    Args:
        items: 待排序条目列表
        mode: 排序模式
            - "dimensionOrder": 按维度 sort_order 排列
            - "customDragOrder": 用户自定义顺序（保持原列表顺序，用上移/下移控制）

    Returns:
        排序后的条目列表
    """
    if mode == "customDragOrder":
        # 列表顺序即用户排列顺序，直接返回
        return list(items)

    # dimensionOrder：按置顶后顺序（gender→ethnicity→height→body...→camera）
    dim_order = {
        "gender": 0, "ethnicity": 1, "height": 2,
        "body": 3, "face": 4, "top": 5, "bottom": 6, "outfit": 7,
        "shoes": 8, "accessories": 9, "pose": 10, "props": 11,
        "background": 12, "camera": 13,
    }
    return sorted(
        items,
        key=lambda it: dim_order.get(it.module.dimension_key or "", 99),
    )

File: src/engine/assembly.ts
/**
 * 拼装引擎 — 对标 src/engine/assembly.py
 */
import { adaptToModel } from './adapters'
import { applyRules } from './rules'
import { evaluateRules } from './ruleEngine'
import type { AssemblyConfig, Dimension, SelectedItem } from './models'
import { PromptIR, IR_SNAPSHOT_VERSION, type IRSegment } from './models'
import type { EngineRule } from './ruleTypes'

export function assemble(
  selected: SelectedItem[],
  config: AssemblyConfig,
  rules?: EngineRule[],
  dimOrderMap?: Record<string, number>,
): { ir: PromptIR; finalPrompt: string } {
  // need02：传入 rules 即走数据驱动求值（段快照 + 锁定集合由 SelectedItem 映射），否则回退 legacy applyRules
  let filtered: SelectedItem[]
  let warnings: string[]
  let findings: PromptIR['findings'] = []
  if (rules) {
    const segments = selected.map(
      (it): IRSegment => ({
        dimensionKey: it.module.dimensionKey ?? '',
        text: it.module.contentEn,
        weight: it.weightOverride != null ? it.weightOverride : it.module.weight,
        sourceModuleId: it.module.id,
      }),
    )
    const lockedIndexes = new Set<number>()
    selected.forEach((it, i) => { if (it.locked) lockedIndexes.add(i) })
    findings = evaluateRules({ segments, rules, lockedIndexes })
    const removed = new Set<number>()
    for (const f of findings) {
      if (!f.fix) continue
      for (const idx of f.fix.removeIndexes) removed.add(idx)
    }
    filtered = selected.filter((_, i) => !removed.has(i))
    warnings = [...new Set(findings.map((f) => f.message))]
  } else {
    ;[filtered, warnings] = applyRules(selected)
  }
  const ordered = sortByOrder(filtered, config.sortBy, dimOrderMap)
  const ir = new PromptIR(
    ordered.map(
      (it): IRSegment => ({
        dimensionKey: it.module.dimensionKey ?? '',
        text: it.module.contentEn,
        weight: it.weightOverride != null ? it.weightOverride : it.module.weight,
        sourceModuleId: it.module.id,
      }),
    ),
    warnings,
    findings,
    IR_SNAPSHOT_VERSION,
  )
  const finalPrompt = adaptToModel(ir, config.modelProfile, config)
  return { ir, finalPrompt }
}

const DIM_ORDER: Record<string, number> = {
  gender: 0,
  ethnicity: 1,
  height: 2,
  body: 3,
  face: 4,
  top: 5,
  bottom: 6,
  outfit: 7,
  shoes: 8,
  accessories: 9,
  pose: 10,
  props: 11,
  background: 12,
  camera: 13,
}

export function buildDimOrderMap(dims: Dimension[]): Record<string, number> {
  return Object.fromEntries(dims.map((d) => [d.key, d.sortOrder]))
}

export function sortByOrder(items: SelectedItem[], mode: string, dimOrderMap?: Record<string, number>): SelectedItem[] {
  if (mode === 'customDragOrder') return [...items]
  // need04：L1 dimOrderMap（DB sortOrder 唯一真相源）> L2 legacy DIM_ORDER > L3 沉底
  return [...items].sort((a, b) => {
    const ka = a.module.dimensionKey ?? ''
    const kb = b.module.dimensionKey ?? ''
    const oa = dimOrderMap?.[ka] ?? DIM_ORDER[ka] ?? Number.MAX_SAFE_INTEGER
    const ob = dimOrderMap?.[kb] ?? DIM_ORDER[kb] ?? Number.MAX_SAFE_INTEGER
    return oa - ob
  })
}
